src/hashset.c

```diff
@@ -0,0 +1,1705 @@
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/* Hashset
+ * =======
+ *
+ * This is an implementation of an open addressing hash table with cache-line
+ * sized buckets. It's designed for speed and low memory overhead. It provides
+ * lookups using a single memory access in most cases and it provides the
+ * following features:
+ *
+ * - Incremental rehashing using two tables.
+ *
+ * - Stateless iteration using 'scan'.
+ *
+ * - A hash table contains pointer-sized elements rather than key-value entries.
+ *   Using it as a set is strait-forward. Using it as a key-value store requires
+ *   combining key and value in an object and inserting this object into the
+ *   hash table. A callback for fetching the key from within the element is
+ *   provided by the caller when creating the hash table.
+ *
+ * - The element type, key type, hash function and other properties are
+ *   configurable as callbacks in a 'type' structure provided when creating a
+ *   hash table.
+ *
+ * Conventions
+ * -----------
+ *
+ * Functions and types are prefixed by "hashset", macros by "HASHSET". Internal
+ * names don't use the prefix. Internal functions are 'static'.
+ *
+ * Credits
+ * -------
+ *
+ * - The design of the cache-line aware open addressing scheme is inspired by
+ *   tricks used in 'Swiss tables' (Sam Benzaquen, Alkis Evlogimenos, Matt
+ *   Kulukundis, and Roman Perepelitsa et. al.).
+ *
+ * - The incremental rehashing using two tables, though for a chaining hash
+ *   table, was designed by Salvatore Sanfilippo.
+ *
+ * - The original scan algorithm (for a chained hash table) was designed by
+ *   Pieter Noordhuis.
+ *
+ * - The incremental rehashing and the scan algorithm were adapted for the open
+ *   addressing scheme, including the use of linear probing by scan cursor
+ *   increment, by Viktor Söderqvist. */
+#include "hashset.h"
+#include "serverassert.h"
+#include "zmalloc.h"
+#include "mt19937-64.h"
+#include "monotonic.h"
+
+#include <limits.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+/* The default hashing function uses the SipHash implementation in siphash.c. */
+
+uint64_t siphash(const uint8_t *in, const size_t inlen, const uint8_t *k);
+uint64_t siphash_nocase(const uint8_t *in, const size_t inlen, const uint8_t *k);
+
+/* --- Global variables --- */
+
+static uint8_t hash_function_seed[16];
+static hashsetResizePolicy resize_policy = HASHSET_RESIZE_ALLOW;
```

Comment:
Can we use type callbacks or values instead of the globals? We discussed changing this in Redis to make it more modular, but never got around to it. This allows much finer control of the policies as well.

Reply:
Yes, we could. It needs some design work. I wanted to avoid side-stepping too much. Just focusing on replacing dict at the first step.

Follow-up.

Reply:
Yeah, followup is always fine.

Reply:
Follow-up: #1364